src/workerd/io/worker.c++

```diff
@@ -3565,4 +3570,13 @@ kj::Promise<WorkerInterface::CustomEvent::Result>
   return inner->customEvent(kj::mv(event));
 }
 
+int32_t Worker::getNextPromiseContextTag() const {
+  int32_t expected = kMaxSmi + 1;
+  __atomic_thread_fence(__ATOMIC_RELAXED);
+  int32_t tag = __atomic_fetch_add(&impl->promiseContextTag, 1, __ATOMIC_RELAXED);
```

Comment:
It's possible that two threads could execute this line before either one executed the follow line, causing the wrap-around point to be missed.

It might be easier to take the result modulo the limit value, so that you only ever need to do increments on the counter value.